Move administering permission tests to `test_channel_permissions.py`.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
This PR moves tests for checking channel administering permission based on `can_administer_channel_group`
setting to `test_channel_permissions.py`.

 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

shubham-padia commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -746,3 +748,241 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+        shiva = self.example_user("shiva")
+        hamlet = self.example_user("hamlet")
+        prospero = self.example_user("prospero")
+        polonius = self.example_user("polonius")
```
Can we give them names such as admin, owner, moderator, etc, depending on the use of that user in this test. It will make keeping track of why a user is used at a certain place without spending time thinking that iago is an admin.

sahil839 replied:
Updated this.

shubham-padia commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -746,3 +748,241 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+                )
+            else:
+                setattr(stream, property_name, old_value)
+                stream.save(update_fields=[property_name])
```
Can the above two if else blocks be combined? Will make it more readable.

sahil839 replied:
We need to call `do_change_stream_group_based_setting` to handle anonymous groups, so we cannot call `setattr` directly.

shubham-padia commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -746,3 +748,241 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+                )
+                do_change_stream_group_based_setting(
+                    stream, "can_subscribe_group", nobody_group, acting_user=iago
+                )
```
Can we extract the private stream testing in it's own function somehow. Even if it's used once, it  will make it easier to read the whole thing I think.

sahil839 replied:
Do you mean we need to add a function in the same scope as `check_channel_property_update`? I don't think doing that helps in readability and defining it outside would mean passing too many parameters.

shubham-padia reviewed (commented):
The logic looks sound, but the functions are a bit hard to read. Left some comments + I think this can use some more comments spread out around the function to make it a bit easier to read.

sahil839 commented:
I was thinking about having separate tests for group permissions. That would help in avoiding many if-else conditions and improving readability. My initial concern to do that was we need to repeat the checks for system group, user-defined group and anonymous group, but I think that would be fine overall in terms of readability and maybe we can have separate test for private stream.

Also, another way to avoid repeated checks for setting set to different values could be something like this below. Maybe a small comment for above each `dict` definition might help.
```py
anonymous_group_dict = UserGroupMembersData(
    direct_members=[prospero.id, guest.id], direct_subgroups=[]
)
group_permission_check_list = [
    dict(
        group=nobody_group, failing_users=[moderator], passing_users=[admin],
    ),
    dict(
        group=moderators_group, failing_users=[hamlet], passing_users=[moderator],
    ),
    dict(
        group=hamletcharacters_group, failing_users=[moderator], passing_users=[hamlet]
    ),
    dict(
        group=anonymous_group_dict, failing_users=[hamlet, moderator], passing_users=[prospero]
    )
]

for test_config in group_permission_check_list:
    do_change_stream_group_based_setting(stream, "can_administer_channel_group", test_config["group"], acting_user=admin)
    for user in test_config["failing_users"]:
        check_channel_property_update(user, allow_fail=True)
    for user in test_config["passing_users"]:
        check_channel_property_update(user)
```

sahil839 commented:
@shubham-padia can you have a look again? I have updated the PR to use the dict approach mentioned above and have also added separate tests for group permissions and stream privacy.

This adds some lines of code but makes the code readable and we would not need to add test for checking permission to update the setting for any new stream group setting.

timabbott commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -746,3 +755,506 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+        default_error_msg = "You do not have permission to administer this channel."
+
+        def check_channel_property_update(
+            user: UserProfile, allow_fail: bool = False, error_msg: str = default_error_msg
```
`s/allow_fail/expect_fail/` would be a better name. I think you probably want just a single parameter, like `expected_error: str`, rather than two parameters that are basically duplicate. We use that sort of pattern in `test_classes` a few places I think.

timabbott commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -746,3 +755,506 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+        # is changed from private to public.
+        if not stream.invite_only and not stream.history_public_to_subscribers:
+            stream.history_public_to_subscribers = True
+            stream.save(update_fields=["history_public_to_subscribers"])
```
I really don't like this model for doing tests -- the leakage risks that some other state remains wrong are considerable, when we could just let Django test system's transactions handle it. It's much better to do each case in a separate test function for this reason.

timabbott commented:
Overall this is probably helpful but we should be mindful of (1) complex test setup, while less repetitive, is much more vulnerable to bugs where the test isn't testing what we think it is and permits us to break the feature, and (2) Checking permissions for an array of user types and an array of settings like this in loops can result in a slow test suite -- especially if we're not careful about avoiding doing expensive calls in every iteration of the loop that is only needed in some parts.
